### PR TITLE
[c++] Reorganize directory to separate public and private headers

### PR DESCRIPTION
**Pull request by nguyenv**
**Issue and/or context:**

#1300 

This is one of the several changes made in #1259 that will be broken off into smaller PRs.

This PR sets up the new directory structure such that, in the future, all public headers are contained in `src/tiledbsoma`. Although for now, we make all headers public as to not break the Python and R APIs. This allows us to merge in code with smaller PRs instead of all at once. After the Python and R bindings are refactored using only the public C++ API, the headers not residing in `src/tiledbsoma` will be made internal only.

### CODE REVIEW

File: libtiledbsoma/src/cli/arrow_adapter.h [hunk 1]
```diff
@@ -1,7 +1,9 @@
 #ifndef ARROW_ADAPTER_H
 #define ARROW_ADAPTER_H
 
-#include <tiledbsoma/tiledbsoma>
+#include <tiledb/tiledb>
+#include "../utils/column_buffer.h"
+#include "../utils/logger.h"
 #ifndef ARROW_SCHEMA_AND_ARRAY_DEFINED
 #include "carrow.h"
 #endif
```
Comment (eddelbuettel) [inline libtiledbsoma/src/cli/arrow_adapter.h:9]: I have learned more recently from the Arrow docs that "they" use header guard `ARROW_C_DATA_INTERFACE`.  The idea (by them) is if one protects with that one should be fine.  We could try that, possibly inside `carrow.h`.

File: libtiledbsoma/src/soma/managed_query.cc [hunk 2]
```diff
@@ -90,6 +90,10 @@ void ManagedQuery::select_columns(
     }
 }
 
+void ManagedQuery::set_column(std::shared_ptr<ColumnBuffer> column_buffer) {
+    column_buffer.get()->attach(*query_);
+}
+
```
Comment (eddelbuettel) [inline libtiledbsoma/src/soma/managed_query.cc:96]: Why do we need this now?  This was previously covered from `ColumnBuffer`.

File: libtiledbsoma/src/soma/managed_query.h [hunk 3]
```diff
@@ -85,6 +83,13 @@ class ManagedQuery {
     void select_columns(
         const std::vector<std::string>& names, bool if_not_empty = false);
 
+    /**
+     * @brief Set column data for write query.
+     *
+     * @param column_buffer Column data
+     */
+    void set_column(std::shared_ptr<ColumnBuffer>);
```
Comment (eddelbuettel) [inline libtiledbsoma/src/soma/managed_query.h:91]: As above

File: libtiledbsoma/src/utils/logger.h [hunk 4]
```diff
@@ -54,7 +54,7 @@
 #if !defined(R_BUILD)
 #include <spdlog/spdlog.h>
 #else
-#include "tiledbsoma/spdlog_with_R.h"
+#include "spdlog_with_R.h"
```
Comment (eddelbuettel) [inline libtiledbsoma/src/utils/logger.h:57]: This is disappearing anyway in the parallel PR / branch I mentioned.

File: libtiledbsoma/src/utils/stats.h [hunk 5]
```diff
@@ -39,10 +39,10 @@
 
 namespace tiledbsoma::stats {
 
-__attribute__((visibility("default"))) void enable();
-__attribute__((visibility("default"))) void disable();
-__attribute__((visibility("default"))) void reset();
-__attribute__((visibility("default"))) std::string dump();
```
Comment (eddelbuettel) [inline libtiledbsoma/src/utils/stats.h:45]: Thanks for cleaning that up.

**Review by eddelbuettel — CHANGES_REQUESTED** [review]
A couple of small changes within which are not critical.

However, it does not build for me (for R) on a checkout.  I can probably try to help.

File: libtiledbsoma/src/soma/managed_query.cc (hunk 2 above)
Comment (nguyenv) [reply libtiledbsoma/src/soma/managed_query.cc:96]: Didn't mean to add that into this PR. I added this method for array writes.

**Comment by nguyenv** [comment]
I'll take a look at that. I had thought it passed CI on R before I openned this PR but obviously was mistaken.

**Review by eddelbuettel — APPROVED** [review]
This looks good.  My one open comment may still apply -- we could try to use the actual Arrow define guard value.  But we can also leave that for another time.

File: apis/r/libtiledbsoma.tar.gz
Comment (teo-tsirpanis) [inline apis/r/libtiledbsoma.tar.gz:1]: What is this file?

Comment (eddelbuettel) [reply apis/r/libtiledbsoma.tar.gz:1]: Long story.  R package builds needs a _complete_ file tree _relative to where they start in the file system_,  This tends to get put into a tar.gz.  This repo is special as it has (from the top) C++, Python, and R.  Now, _currently_ R needs this tarball to build as it depends on the C++ library.  This allows for example builds at external builder such as [this creating us binaries](https://tiledb-inc.r-universe.dev/tiledbsoma).

We have [parallel work ongoing](https://github.com/single-cell-data/TileDB-SOMA/pull/1280) to allow for that library to be an external resource, but that build currently fails on macOS and is therefore not merged or progressing much.

**Comment by johnkerl** [comment]
For the record here are before-and-afters with `scripts/bld --prefix=build-old` and `scripts/bld --prefix=build-new`, respectively

```
$ find build-old -follow -type f
build-old/include/span/span.hpp
build-old/include/thread_pool/producer_consumer_queue.h
build-old/include/thread_pool/status.h
build-old/include/thread_pool/thread_pool.h
build-old/include/tiledbsoma/array_buffers.h
build-old/include/tiledbsoma/arrow_adapter.h
build-old/include/tiledbsoma/carrow.h
build-old/include/tiledbsoma/column_buffer.h
build-old/include/tiledbsoma/common.h
build-old/include/tiledbsoma/logger_public.h
build-old/include/tiledbsoma/managed_query.h
build-old/include/tiledbsoma/soma_array_reader.h
build-old/include/tiledbsoma/spdlog_with_R.h
build-old/include/tiledbsoma/stats.h
build-old/include/tiledbsoma/tiledbsoma
build-old/include/tiledbsoma/tiledbsoma_export.h
build-old/include/tiledbsoma/util.h
build-old/include/tiledbsoma/version.h
build-old/lib/cmake/tiledbsoma/TileDBSomaConfig.cmake
build-old/lib/cmake/tiledbsoma/TileDBSomaTargets-release.cmake
build-old/lib/cmake/tiledbsoma/TileDBSomaTargets.cmake
build-old/lib/libtiledb.so.2.15
build-old/lib/libtiledbsoma.so
build-old/lib/pkgconfig/tiledbsoma.pc
build-old/bin/tdbsoma
```

```
$ find build-new -follow -type f
build-new/include/cli/arrow_adapter.h
build-new/include/cli/carrow.h
build-new/include/cli/cli.cc
build-new/include/soma/managed_query.cc
build-new/include/soma/managed_query.h
build-new/include/soma/soma_array.cc
build-new/include/soma/soma_array.h
build-new/include/tiledbsoma/logger_public.h
build-new/include/tiledbsoma/tiledbsoma
build-new/include/tiledbsoma/tiledbsoma_export.h
build-new/include/utils/array_buffers.h
build-new/include/utils/column_buffer.cc
build-new/include/utils/column_buffer.h
build-new/include/utils/common.h
build-new/include/utils/logger.cc
build-new/include/utils/logger.h
build-new/include/utils/stats.cc
build-new/include/utils/stats.h
build-new/include/utils/util.cc
build-new/include/utils/util.h
build-new/include/utils/version.cc
build-new/include/utils/version.h
build-new/lib/cmake/tiledbsoma/TileDBSomaConfig.cmake
build-new/lib/cmake/tiledbsoma/TileDBSomaTargets-release.cmake
build-new/lib/cmake/tiledbsoma/TileDBSomaTargets.cmake
build-new/lib/libtiledb.so.2.15
build-new/lib/libtiledbsoma.so
build-new/lib/pkgconfig/tiledbsoma.pc
build-new/bin/tdbsoma
```

**Pull request merged**